Sources/NIOSSH/Connection State Machine/SSHConnectionStateMachine.swift

```diff
@@ -57,11 +57,21 @@ struct SSHConnectionStateMachine {
         case sentDisconnect(SSHConnectionRole)
     }
 
+    class Attributes {
```

Comment:
Ah, I see: it's not a `struct` because we need some action at a distance to store the value.

I don't love that. I wonder if we can refactor the code a bit. We have a few options.

The first option is that we could force the user auth delegate to tell us which username they're accepting. That's kind of tricky to do and a bit awkward.

The second option is that we could store a `class` only for the lifetime of the key exchange, and use that. Then reify this into a struct.

The third option is that we could use a promise to resolve the username. I don't love that much, but it's do-able without being too painful.

The fourth option is to change the return value convention here somewhat so that we can also resolve a side-effect. Do you have any thoughts here?

Reply:
Hey Cory, thanks for the quick follow-up! I concur that it's sub-optimal. I merely updated the PR so you could leave your feedback, but will take it from here.

I don't think option 1 and 3 is a great API. It's going to be an awkward API to use for sure. One good option - in my eyes - is sending an event on the channel that indicates which user was authenticated. This leaves catching it to the implementations like Citadel.

Reply:
I'm open to sending an event on the channel.